### ISSUE TITLE
[Enhancement]: Improve Anchor Tag Accessibility in WooCommerce

### ISSUE THREAD

**Issue by shkarinn** [woothemes/woocommerce#48388]
![brave_fBkCGsg7AP](https://github.com/woocommerce/woocommerce/assets/29929365/a8f77fdf-b549-4442-b372-3561a59439cd)

**Issue:** Google PageSpeed Insights reports a deduction due to anchor tags lacking title or aria-label attributes, impacting accessibility for users relying on assistive technologies.

Guideline Reference: [Deque University link-name rule](https://dequeuniversity.com/rules/axe/4.9/link-name) emphasizes the need for discernible, unique, and focusable link text to improve navigation for screen reader users.

### Proposed Solution:
Add aria-label or title Attributes:
Ensure all anchor tags include meaningful aria-label or title attributes for better context and accessibility.

### Example:
**Current Code:**

`<a href="/example">Read More</a>`

**Proposed Code:**

`<a href="/example" aria-label="Read more about Example Product">Read More</a>`

Benefits:

- Improved Accessibility: Better experience for users with assistive technologies.
- Enhanced SEO: Improved semantic markup.
- Compliance: Aligns with accessibility standards.

### Additional context

This is the part of wc-template-functions.php file need to fix

```
function wc_get_gallery_image_html( $attachment_id, $main_image = false ) {
	$flexslider        = (bool) apply_filters( 'woocommerce_single_product_flexslider_enabled', get_theme_support( 'wc-product-gallery-slider' ) );
	$gallery_thumbnail = wc_get_image_size( 'gallery_thumbnail' );
	$thumbnail_size    = apply_filters( 'woocommerce_gallery_thumbnail_size', array( $gallery_thumbnail['width'], $gallery_thumbnail['height'] ) );
	$image_size        = apply_filters( 'woocommerce_gallery_image_size', $flexslider || $main_image ? 'woocommerce_single' : $thumbnail_size );
	$full_size         = apply_filters( 'woocommerce_gallery_full_size', apply_filters( 'woocommerce_product_thumbnails_large_size', 'full' ) );
	$thumbnail_src     = wp_get_attachment_image_src( $attachment_id, $thumbnail_size );
	$full_src          = wp_get_attachment_image_src( $attachment_id, $full_size );
	$alt_text          = trim( wp_strip_all_tags( get_post_meta( $attachment_id, '_wp_attachment_image_alt', true ) ) );
	$image             = wp_get_attachment_image(
		$attachment_id,
		$image_size,
		false,
		apply_filters(
			'woocommerce_gallery_image_html_attachment_image_params',
			array(
				'title'                   => _wp_specialchars( get_post_field( 'post_title', $attachment_id ), ENT_QUOTES, 'UTF-8', true ),
				'data-caption'            => _wp_specialchars( get_post_field( 'post_excerpt', $attachment_id ), ENT_QUOTES, 'UTF-8', true ),
				'data-src'                => esc_url( $full_src[0] ),
				'data-large_image'        => esc_url( $full_src[0] ),
				'data-large_image_width'  => esc_attr( $full_src[1] ),
				'data-large_image_height' => esc_attr( $full_src[2] ),
				'class'                   => esc_attr( $main_image ? 'wp-post-image' : '' ),
			),
			$attachment_id,
			$image_size,
			$main_image
		)
	);

	return '<div data-thumb="' . esc_url( $thumbnail_src[0] ) . '" data-thumb-alt="' . esc_attr( $alt_text ) . '" class="woocommerce-product-gallery__image"><a href="' . esc_url( $full_src[0] ) . '">' . $image . '</a></div>';
}
```

**Comment by Aljullu** [comment]
Thanks for opening this issue, @shkarinn!

You can easily fix this issue adding an `alt` text to the product images you upload. You can do so going to Media > Library, clicking on the product image that is causing the issue and adding an Alternative Text:

![image](https://github.com/woocommerce/woocommerce/assets/3616980/22d01b21-96f8-4a6a-9c5e-d832a41e0ec2)

That's definitely the solution I would recommend you to try.

Having said that, I agree that WooCommerce could do a better shop and provide a default alt text when it's missing.

**Steps to reproduce**

1. Create a product and upload some images for it. Don't set the `alt` text.
2. Go to the product template in the frontend and using your browser devtools _Inspector_, check `.woocommerce-product-gallery__image > a > img`.
3. Notice `alt` is missing, even though it's clearly not a decorative image.

Expected behavior: while the best approach is that store owners write their own `alt` texts, WooCommerce could provide a default when it's missing, something like "[product] image 1", "[product] image 2", etc.

**Comment by shkarinn** [comment]
> You can easily fix this issue adding an `alt` text to the product images you upload. You can do so going to Media > Library, clicking on the product image that is causing the issue and adding an Alternative Text:

Thanks for advice, in this case I will be prefere auto generate all alt tags in separate function, because I have over 1k items. But I didn't found filter for $alt_text. This filter will be very helpful for auto generation alt text for img.

**Comment by shkarinn** [comment]
> **Steps to reproduce**
> 
> 1. Create a product and upload some images for it. Don't set the `alt` text.
> 2. Go to the product template in the frontend and using your browser devtools _Inspector_, check `.woocommerce-product-gallery__image > a > img`.
> 3. Notice `alt` is missing, even though it's clearly not a decorative image.

Thanks, I add alt text for image, but Google Page Speed found new accessibility problem.

![brave_c7MXviVGP2](https://github.com/woocommerce/woocommerce/assets/29929365/dfa9f040-48e2-4703-b95e-2efb89009dac)

**Comment by Aljullu** [comment]
> Thanks, I add alt text for image, but Google Page Speed found new accessibility problem.

Thanks for mentioning it, @shkarinn! It looks like we already have an issue for it: https://github.com/woocommerce/woocommerce/issues/44478, I just provided some more info there so we (or somebody from the community) can hopefully take a look soon.

**Comment by amesplant** [comment]
I am looking into this issue this week. This particular issue seems to be less about the image alt tag than it does for how the actual link should be read by a screen reader. I agree that the link should include more text available to a screen reader than the "Read More". I will look at implementing the product title into the link text, specifically for screen readers.

**Comment by amesplant** [comment]
@rrennick @Aljullu 
It looks like my colleague @Manussakis is looking into the issue re: the image's role:
https://github.com/woocommerce/woocommerce/issues/44478

While adding the `alt` tag to an image is a "fix", could we consider adding a title field to each of the product images? Or minimally a note to remind users to add alt tags to each product image?

**Comment by Aljullu** [comment]
Hey @amesplant and @Manussakis, thanks for looking into this!

> While adding the `alt` tag to an image is a "fix", could we consider adding a title field to each of the product images?

When uploading an image to WordPress, the title is set to the file name, if I'm not wrong. Is that what you are referring? I think in this case, using the title is unlikely to be more beneficial than using the alt text, no?

![imatge](https://github.com/user-attachments/assets/fe3710bb-a6a7-4f30-a294-2b71f53ddf0c)

> Or minimally a note to remind users to add alt tags to each product image?

Where would this note be displayed? Heads-up that this UI is mantained by WordPress, even though we can introduce changes, I think this kind of modifications would be better done upstream. What do you think?

(I might be completely misunderstanding your solution, though, sorry if that's the case! :sweat_smile: )

**Comment by amesplant** [comment]
@Aljullu 
I meant to this area, thinking that once the user adds the product image, they would see a Title field that may be more intuitive for them to add vs. the alt tag in the image upload itself. I think a lot of content users don't necessarily fully understand what the `alt` tag provides to users.
![image](https://github.com/user-attachments/assets/edcef75f-c5b4-45a6-ab6a-042879eaa394)

If this is too big of an ask, then I think that this entire ticket is really a "non-issue", since the current solution exists by adding the appropriate `alt` tag to the image.

**Comment by Aljullu** [comment]
Ah, thanks for answering, @amesplant. While it makes sense, I think it's unlikely we will change the UI of the classic product editor, as we are working on a new one. However, as we are working on the new product editor, it might be a good opportunity to add this kind of improvements.

I will ping @louwie17 and @jarekmorawski as they have more experience on it.

_tl;dr_ some users forget to add the alt text when uploading product images. @amesplant suggested updating the product editor UI to make it easier to update the alt text. I tested, and in the new product editor there is no easy way to edit the alt text of already-selected images:

![imatge](https://github.com/user-attachments/assets/324e6c70-8969-4245-9455-677c3c5867f1)

So the questions are:

1. Has there been any explorations on making it easier to modify the alt text of images directly from the product editor?
2. What would you think about adding some kind of warning or `!` sign to images which are missing the alt text?

Images missing their alt text is bad for accessibility and for SEO, so ideally we should notify store owners when that happens. What do you think?

**Comment by louwie17** [comment]
Thanks for tagging us @Aljullu, I like the overall idea of making this easier. I will answer your two follow up questions below:

> 1. Has there been any explorations on making it easier to modify the alt text of images directly from the product editor?

No, not that I am aware. In generally for the Images block we tried to follow the WP Images/Gallery block. I noticed in the post editor the alt text is pretty explicitly shown in the block toolbar. We don't use the block toolbar in the product form, but maybe we can add an option to the image toolbar.
![Screenshot 2024-07-26 at 2 16 57 PM](https://github.com/user-attachments/assets/a375a133-2c4d-4881-bd45-b1ae2f888071)

> 2. What would you think about adding some kind of warning or `!` sign to images which are missing the alt text?

Not a bad idea! I will tag @j111q also as she is the primary designer now for the new product form.

**Comment by amesplant** [comment]
@louwie17 cc @vikrampm1 should we move this into an "on hold" status or back to "to do" until we have a clear direction on how to address adding a descriptive text field for images?

**Comment by vikrampm1** [comment]
Moving this back to "To Do" as this needs further clarification before making any progress.

**Comment by vikrampm1** [comment]
@j111q can you able to review the above [comments](https://github.com/woocommerce/woocommerce/issues/48388#issuecomment-2252643718) and help us how to best proceed forward?

**Comment by j111q** [comment]
Hi 👋  to me, I don't think we should prompt the user with a warning because we generally never do that for anything except the product name - like even if the product doesn't have a price, we don't flag anything to the user. 

I think I'd vote for auto-generating the Alt Text using `[product name] Image 1`, `[product name] Image 2`. Because usually product names are really really descriptive (for SEO purposes), and so I think they'd be a good start as an alt text description, and ideally users would take that initial description and just tweak it to be more specific to the image. 

We could also consider surfacing it under the main Product image, so that the user is aware of this field and can see what has been auto-generated, something like this:

![image](https://github.com/user-attachments/assets/bfbaa1f5-b09c-448f-8595-544cedd1bd6c)

I hesitate to use the actual image file name because usually it's something the merchant assumes the shopper will never see -- so it might contain the SKU number or something like that.

Outside of the product editor, we should add something like an inbox note guiding people on accessibility best practices so they understand the value of alt text -- I know @jamesckemp is looking at this area, maybe he might have some thoughts here.

**Comment by vikrampm1** [comment]
cc @Manussakis

**Comment by jamesckemp** [comment]
@j111q I like all of your suggestions on this!
